### ISSUE TITLE
Option -bed_def lc seems to break reproducibility depending on resolution

### ISSUE THREAD

**Issue by talbrecht** [pism/pism#377]
Reproducibility of PISM simulations is important in terms of parameter sensitivity. One would expect for exactly same parameters (including resolution and cpu) exactly same results...

We learned from the PETSc community that CPU is a model parameter and though we have this issue solved by setting intel parameter flags` -fp-model precise` for PETSc configuration, since it is very likely related to SSA calculations (no energy scheme nor I/O). 

I ran a [simple test for a simplified searise-antarctica example](http://www.pik-potsdam.de/~albrecht/pism_repro/exp.sh) (constant-in-time forcing, no PDD) with the dev code over 1000 model years and different resolutions and found partly reroducibility but in some cases divergence already after about 100 model years. [The plots](http://www.pik-potsdam.de/~albrecht/pism_repro/plots/) reveal some dependence on resolution (4/5 reproduced for 50km and 30km, 5/5 for 20km and 0/5 for 15km resolution)... I became aware of this problem when switching my whole paleo ensemble analysis from 20 to 15km with quite some impact... The input field is on a 5km grid (cdo remapcon instead of PISM internal regridding does not solve the problem!).

Also for local 1 CPU simulations on my laptop (using mpich) I get quite some divergence (in all cases for 50km, in 1/5 for 30km and in no case for 20km!)... @matthiasmengel is checking on the LRZ cluster...

**Comment by JoSutter** [comment]
Hi Thorsten,

it doesn't surprise me at all that the model results are resolution dependent, especially over such a wide range of resolution. I think that is to be expected as for example channeling of ice flow will look different. From your plots the simulation seems to be only slightly affected by cpu number, this irritates me as well, but maybe this is rooted in some precision point issue?

cheers,

Johannes

**Comment by JoSutter** [comment]
sorry for the name-spelling faux pas :)

**Comment by talbrecht** [comment]
Johannes,
different solutions for different resolutions or different CPU doesn't surprise an more. But different solutions for exactly the same resolution and CPU (in some cases) is an issue, at least it adds some kind of model-internal uncertainty, which can be large for strong forcing (e.g. for [deglaciation simulations](http://www.pik-potsdam.de/~albrecht/pism_repro/old_plots_long_runs/deglac_runs_50-52_8cpu.pdf) )...

**Comment by tkleiner** [comment]
I am actually not surprised that resolution matters. In e.g. Golledge et al., 2015 (doi:10.1038/nature15706) different enhancement factors have been applied to account for "missing physics" on different length scales.  The resolution dependency of ice flow models has also been shown in many of the MIPs in the past. If everything would be "right" on the 50km scale one would not need to waste CPU time for higher resolution runs.  

The 1cpu and 8cpu simulations using 50km resolution should be very similar, as shown for the sea-level equivalent. Not sure, why the horizontal velocity differs so much in between the 1cpu and 8cpu case of the same resolution. This would be a problem. I will check this on our computer.
 
P.S.:  I was not aware of `-fp-model precise`. Thanks for this.

**Comment by JoSutter** [comment]
Torsten,

if the numbers in your deglaciation figure (50,51,52) are labelling resolution (km) then I'm also surprised about the magnitude of the changes.

**Comment by talbrecht** [comment]
@tkleiner: The intel compiler flag option was a result of some discussions with Stefan Petri here at PIK more than half a year ago... I think Constantine wanted to add it to the installation manual. 
But these days I ran all the reproducibility tests with 20km resolution (and Matthias now with 10km), such that this issue just didn't show up... 
@JoSutter: no the number is just some label for the number of the simulation, all settings are exactly the same otherwise...

**Comment by tkleiner** [comment]
Could you try without "skip=5" (`exp.sh`)? I decided not to use this any more for coarse resolution simulations 40km to 10km.  I found that simulations became very unstable using skip for PISM versions since 0.7.2 (Fixed bug in enthalpy scheme).

**Comment by talbrecht** [comment]
Thanks Thomas for looking into it. Yes, I did already, this changed the solution as such (also a model parameter!), but it doesn't change the fact that 20km (10km) is perfectly reproducible and 15km (...) is not at all...

**Comment by ckhroulev** [comment]
@talbrecht Could you please remove all hard-wired file and directory paths from `exp.sh`? This would make it much easier for me to run it. Thanks!

Also, what is "`noexf`"?

**Comment by ckhroulev** [comment]
Regarding `-fp-model precise`: it is mentioned in [`INSTALL.md`](https://github.com/pism/pism/blob/dev/INSTALL.md) (the `dev` branch).

**Comment by tkleiner** [comment]
@talbrecht after recompiling PETSc and PISM using `-fp-model precise` I run `exp.sh` with two different configurations for the bed deformation (-bed_def lc, iso) for the 15km set-up only. I think the problem is within -bed_def lc. See figures 
![bediso_15km](https://cloud.githubusercontent.com/assets/5449934/23696079/7e00f5ee-03e2-11e7-9585-d6d125ab1fc4.png)
![bedlc_15km](https://cloud.githubusercontent.com/assets/5449934/23696080/7e01a7c8-03e2-11e7-9451-a060136b6f8a.png)


Sorry for the wrong scaling and units for `max_hor_vel`.  For simplicity I have used `pism-dev/examples/std-greenland/tsshow.py` to plot.  

There was already a hint in #349 where the model crashed at unpredictable time steps with -bed_def lc running the same set-up again and again. 

Do I need to recompile all dependencies (GSL, FFTW,...) using `-fp-model precise`?

**Comment by talbrecht** [comment]
@tkleiner: thanks for checking for 15km resolution. I have made the same observation for switching of `-bed_def lc`: all simulations for 12, 15, 20, 30 and 50km become reproducible in a test of 9 runs each... So we know where to look at, right?!

I have compiled only PETSc with impi option `-fp-model precise`, but for consistency you may be right!

@ckhroulev: "`noexf`" means no output of extra files, I just wanted to check if this may have an influence. And sorry for the misleading initial title of the issue and the overloaded run script. I added a lite version of the [run file `exp_lite.sh`](http://www.pik-potsdam.de/~albrecht/pism_repro/exp_lite.sh).

If using `-bed_def lc` 20km still seems most robust (in all simulations exactly same results) and 15km reveals most divergence. For other tested resolutions (50km, 30km and 12km) non-reproducible simulations occur only rarely (added some more plots)...

**Comment by matthiasmengel** [comment]
Just for reference:
On the LRZ cluster, I get different solutions for resolutions from 12km to 50km, tested on 8 as well as on 28 cores. `-bed_def lc` always on.

**Comment by talbrecht** [comment]
@ckhroulev: I tested the latest dev commit e9d2d1f for 5 different resolutions (9 runs) and all of them showed perfect reproducibility, also for 15km... Yeahaw! Thank you so much!

**Comment by tkleiner** [comment]
@ckhroulev: Many thanks! I would like to ask for a new stable 0.7.4.

**Comment by talbrecht** [comment]
There're a lots of new code related to this issue. What was the key in solving the problem?

Short question: I see the new field viscous_bed_displacement saved to file if option -bed_def lc set, but failing restart (opts.type == INIT_RESTART) when not in input file. Would it make sense here to create a new viscous_bed_displacement variable (e.g. after -no_mass run) as for (opts.type == INIT_BOOTSTRAP)?

**Comment by ckhroulev** [comment]
> There're a lots of new code related to this issue. What was the key in solving the problem?

I'm happy to see that the fix for #370 seems to have taken care of this issue as well. I would appreciate it if you (i.e. all of you) could do some runs and let me know if the new code behaves as expected. More testing is always good; it would be especially helpful if you could try some runs that re-start the `-bed_def lc` model with re-gridding.

The key to solving #370 was saving the viscous bed displacement *on the extended grid used by the spectral collocation method*. While in theory the viscous bed displacement can be recovered using ice thickness and bed uplift maps (see section 4 and equation 16 of "Fast computation of a viscoelastic deformable Earth model for ice-sheet simulations" by Bueler, Lingle, and Brown), in practice this cannot be done **exactly**, for at least two reasons:

  1) the uplift map saved by PISM is an *approximation* computed using finite differences, and
  2) while we can assume that ice thickness is zero outside the physical domain, we **cannot** assume that the resulting uplift map is zero at (and beyond) the boundary of the physical domain. Before the fix PISM saved the uplift map restricted to the physical domain, effectively setting it to zero outside the domain every time PISM is re-started.

There is also the easy-to-fix-but-still-important issue of using consistent values of *t* in equation 17 when applying the displacement modification described in the paper. (I.e. before the fix this correction was not applied consistently after re-starting).

> Short question:

In my experience short questions tend to require long answers. :-)

> I see the new field `viscous_bed_displacement` saved to file if option `-bed_def lc` set, but failing restart `(opts.type == INIT_RESTART)` when not in input file.

This is intentional.

> Would it make sense here to create a new viscous_bed_displacement variable (e.g. after -no_mass run) as for (opts.type == INIT_BOOTSTRAP)?

I'm not quite sure what you mean by "create a new `viscous_bed_displacement` variable (e.g. after `-no_mass` run) as for `(opts.type == INIT_BOOTSTRAP)`".

The command-line option "`-i input.nc`" (the `INIT_RESTART` case) is for re-starting PISM with the same settings as in the run used to produce `input.nc`. If you need to "add more physics", use the "`-bootstrap -i input.nc -regrid_file ...`". When bootstrapping PISM will read in the bed topography and the uplift and compute the corresponding viscous bed displacement.

The problem here is that the bed displacement field does **not** directly correspond to an *observed* physical quantity. It is specific to the bed deformation model implemented in PISM and so PISM should not get it from the user. There is no reasonable default value we could use for it; the default of zero would make some sense, but in my opinion it is not a "reasonable" one.

The user needs to be aware of the fact that initializing the Lingle-Clark bed deformation model requires **both** the bed topography and the uplift map. We allow bootstrapping without an uplift map (using the default of zero), but unlike the viscous bed displacement the uplift can both be measured and interpreted -- the default of zero corresponds to the assumption that the current load is in equilibrium with the Earth.

To summarize: *I want the user to be aware of the fact that if you omit the uplift map, you assume that the current load is in equilibrium with the Earth.* PISM should not make this assumption on the user's behalf.

Please let me know if this helps.

**Comment by ckhroulev** [comment]
@tkleiner Re. 0.7.4: yes, that's the plan. I'm hoping to release 0.8 soon but I realize that ongoing projects using 0.7 will benefit from this fix.

**Comment by talbrecht** [comment]
Great, I understand! Really good! You mean "exact" with regard to the truncation error of approximated equations, and I am happy already with "exactly" the same results for same settings... I will run some more test using grid refinement..

Thanks for pointing out the initial uplift condition, I am actually using modern uplift rates (by Pippa Whitehouse GIA model calibrated against GPS measurements, personal communication) as initial condition assuming Bedmap2 is not in equilibrium. 

OK, I have learned now that the` -bootstrap -regrid_file` combination is made not only for refinement of the grid but also for adding physics, wasn't aware of this. However, in a SeaRISE-like spin-up with boot phase (with` -bed_def lc`) followed by `-no_mass` phase (without` -bed_def lc`) I need to add inputs of the different phases for the full-physics simulation (uplift and viscous_bed_displacement from boot-output and enthalphy from nomass-output). I'm using nco tools for this, which is a little workaround...

Thank you again Constantine

**Comment by ckhroulev** [comment]
@talbrecht In general, `-bootstrap -i input.nc` is what you use when `input.nc` is "incomplete", i.e. misses some fields needed to re-start the model. This is sometimes the case when you add physics; turning on a bed deformation model is one example.

PS: It should now be possible to refine the grid (combining higher-resolution, more detailed bed topography with bed deformation effects from a previous run) by re-gridding `viscous_bed_displacement` from a coarse-grid run onto the refined grid.

**Comment by talbrecht** [comment]
@ckhroulev: I re-ran the full suite of resolution-dependent tests with the fixed dev code e9d2d1f (I also checked for 386eeea). Unfortunately not all of the cases show reproducibility, see [runscript](http://www.pik-potsdam.de/~albrecht/pism_repro/exp_lite.sh) and [plots](http://www.pik-potsdam.de/~albrecht/pism_repro/plots) for 12, 15, 20, 30 and 50km. I am afraid, that the bug fix did not catch the whole problem... 

Did anyone else run tests with the latest dev? Any further ideas how bed deformation could induce non-determinism? 

Regarding the new regression test, `make test` fails for me 
```
The following tests FAILED:
	 26 - bed_deformation:LC:exact_restartability:bed_def_lc_restart.sh (Failed)
Errors while running CTest
```
Does this mean it is a platform/software issue?

**Comment by talbrecht** [comment]
I extended test/regression/bed_def_lc_restart.sh to run 40 times for a 101x101 grid (15km) and replaced options `-stress_balance none -energy none ` by `-pik -sia_e 2.0 -ssa_e 0.6 -ssa_method fd -stress_balance ssa+sia -pseudo_plastic -pseudo_plastic_q 0.25`. 
On the PIK HPC (1CPU) this test failed in 15 of 40 cases... 
```
Comparing ex.nc and ex-restart.nc
Variable dbdt: delta = 1.192093e-07, tol = 0.000000e+00
Files are different.
```
Interestingly, using value `-pseudo_plastic_q 1.0` (linear) passes in all 40 cases on the PIK HPC. Any ideas?

**Comment by tkleiner** [comment]
@talbrecht: I changed `test/regression/bed_def_lc_restart.sh`according to your run with `-pseudo_plastic -pseudo_plastic_q 0.25` and run the experiment 40 times on 1CPU at AWIs Cray-Cluster computer. The proposed test passed in all cases (see slurm output 
[slurm-493311.out.gz](https://github.com/pism/pism/files/903455/slurm-493311.out.gz)).  

These are the external packages installed for pism-dev (386eeea):
FFTW3 (3.4.3), GSL2 (2.1), HDF5 (1.8.17), NETCDF4 (4.4.0), PNETCDF (1.7.0), PROJ4 (4.9.1), UDUNITS2 (2.2.20), PETSC (3.5.4), CMake (3.6.1), intel.compiler (17.0.1)

I have recompiled FFTW3, GSL2, PETSC and PISM using the compiler options `-fp-model precise -no-fast-transcendentals` according to the recommendations found in https://github.com/pism/pism/blob/dev/INSTALL.md for compilation with the Intel compiler and some input from the FESOM community at AWI.  I am not sure if `-no-fast-transcendentals` is really required and if it would have been enough just to recompile FFTW3 for  `-bed_def lc`.

**Comment by talbrecht** [comment]
@tkleiner : Thanks for checking on AWI's cray. So you think this effect may be rather a result of some inconsistent package configuration than a PISM internal bug?

These are the external packages installed for pism-dev (386eeea) on PIK's IBM NeXtScale Cluster:
FFTW (3.3.4), GSL (1.16), HDF5 (1.8.18), NETCDF4 (4.4.1.1), PROJ4 (4.9.3), UDUNITS2 (2.2.19), PETSC (3.7.5), CMake (3.2.3), compiler: Intel(R) MPI Library 2017 Update 1 for Linux OS

I also reinstalled FFTW(3.3.6), GSL(2.3), PETSc(3.7.5) and UDUNITS(2.2.24) for PISM-dev(386eea) with compiler flags` -fPIC -g -O3 -fp-model precise -no-fast-transcendentals`... Unfortunately about 50% of the test runs fail... In fact, excluding options `-ssa_e 0.6` in fact leads to 0 failed tests... (similar to excluding  `-pseudo_plastic_q`)

**Comment by talbrecht** [comment]
Also tested without `-fPIC -g -O3` and AWI-cray-like configuration, still about 25% failed restart tests...

**Comment by talbrecht** [comment]
Interesting is, that submitted with slurm to the PIK cluster yields 100% passed [bed_def_lc_restart tests](http://www.pik-potsdam.de/~albrecht/pism_repro/bed_def_lc_restart_cluster.sh), also if run for longer than 1000years. Here's my [printenv](http://www.pik-potsdam.de/~albrecht/pism_repro/env_local.txt) for the related [pism_build](http://www.pik-potsdam.de/~albrecht/pism_repro/build_functions.sh). However, if I submit the[ reproducibility tests](http://www.pik-potsdam.de/~albrecht/pism_repro/exp_lite.sh) for the SeaRISE Antarctica setup, I still get unreproducible results in many cases. Any ideas?

**Comment by tkleiner** [comment]
@talbrecht  I run your `exp_lite.sh` for the 15km setup within slurm. At least one out of 40 runs is off. I have no idea what's going on here. A better test (better than plotting all lines on top of each other) would be helpful. 
![slvol_bedlc_15km_small](https://cloud.githubusercontent.com/assets/5449934/25008254/7809137c-2063-11e7-9aa4-916fbd14dc5d.png)
![max_hor_vel_bedlc_15km_small](https://cloud.githubusercontent.com/assets/5449934/25008329/a9c29000-2063-11e7-99f7-6e0256447f4d.png)

**Comment by talbrecht** [comment]
@tkleiner: looks pretty good, I observe much more diversity in the solutions when submitted to the PIK cluster (even with same configurations). How many cpu did you use? And what are your srun options? In contrast to all of my solutions with the same run script, yours show a steady increase in slvol.

PS: You could use /util/nccmp.py to compare output variables as in the` test/regression/bed_def_lc_restart.sh` regression test.

**Comment by talbrecht** [comment]
Thanks @tkleiner, I followed precisely your instructions and ran again the` exp_lite.sh` experiment, with
`export OMP_NUM_THREADS=1` and with "srun reconfigure" in the petsc installation. (Only) one out of 40 is (slightly) off (compare plots for[ new ](http://www.pik-potsdam.de/~albrecht/pism_repro/plots/repro_testrunspik_8cpu_15km_500yrs_newcode_newconfig.pdf)and [old](http://www.pik-potsdam.de/~albrecht/pism_repro/plots/repro_testrunspik_8cpu_15km_500yrs_newcode.pdf) configuration for related software). Same result as in your test, also when using nccmp.py for checking results for topg and dbdt. I'm not sure if OMP_NUM_THREADS has helped or the empty PIK cluster over Easter holidays?! I used "srun -n 8 "cpu on one node. Happy Easter!

**Comment by tkleiner** [comment]
Not sure if one out of 40 is good enough, but what irritates  me more is the fact that your experiments all (12km, ..., 50km) show a decrease in `slvol`, while I see an increase over time for the same experiment! 

Just to be sure not to overlook something please find my version of 
[exp_lite.sh.txt](https://github.com/pism/pism/files/924319/exp_lite.sh.txt) with only very minor edits. 

Also the runs using only 8 cores instead of 36 show an increase in `slvol`. This is some output from `ncdump -h ts_test_15km_1_500yrs.nc`:

```
		:source = "PISM (development 386eeea committed by Constantine Khroulev on 2017-03-27 10:04:47 -0800)\n",
			"CMake 3.6.1.\n",
			"Petsc Release Version 3.5.4, May, 23, 2015 \n",
			"PETSc configure: --known-level1-dcache-size=32768 --known-level1-dcache-linesize=64 --known-level1-dcache-assoc=8 --known-memcmp-ok=1 --known-sizeof-char=1 --known-sizeof-void-p=8 --known-sizeof-short=2 --known-sizeof-int=4 --known-sizeof-long=8 --known-sizeof-long-long=8 --known-sizeof-float=4 --known-sizeof-double=8 --known-sizeof-size_t=8 --known-bits-per-byte=8 --known-sizeof-MPI_Comm=4 --known-sizeof-MPI_Fint=4 --known-mpi-long-double=1 --known-mpi-int64_t=1 --known-mpi-c-double-complex=1 --known-sdot-returns-double=0 --known-snrm2-returns-double=0 --with-valgrind=0 --prefix=/global/AWIsoft/tkleiner/petsc/3.5.4_intel_impi --PETSC_ARCH=linux-intel-impi --with-cc=mpiicc --with-cxx=mpiicpc --with-fc=0 --CXXFLAGS=\"-fp-model precise -no-fast-transcendentals\" --CFLAGS=\"-fp-model precise -no-fast-transcendentals\" --with-debugging=0 --with-shared-libraries=1 --with-blas-lapack-dir=/global/AWIsoft/intel/2017/compilers_and_libraries_2017.1.132/linux/mkl/lib/intel64 --with-batch=1 --with-mpiexec=srun --known-mpi-shared-libraries=1\n",
			"Intel(R) MPI Library 2017 Update 1 for Linux* OS\n",
			"NetCDF 4.4.0 of Feb 27 2017 08:29:34 $.\n",
			"FFTW fftw-3.3.4.\n",
			"GSL 2.1.\n",
			"PROJ.4 Rel. 4.9.1, 04 March 2015.\n",
			"" ;
```

**Comment by talbrecht** [comment]
Good point, I think I updated in the meanwhile the `exp_lite.sh` run script, such that the loop reads the `$bootfile` and does not overwrite `$INNAME` as used also for the atmosphere forcing (see [exp_lite_new.sh](http://www.pik-potsdam.de/~albrecht/pism_repro/exp_lite_new.sh)). However, this does not explain the different trend in the test results... 

I am running some more tests with 32 cores (2x16) now. This is my `ncdump -h` output: 
```
:source = "PISM (development 386eeea committed by Constantine Khroulev on 2017-03-27 10:04:47 -0800)\n",
			"CMake 3.2.3.\n",
			"Petsc Release Version 3.7.5, unknown \n",
			"PETSc configure: --known-level1-dcache-size=32768 --known-level1-dcache-linesize=64 --known-level1-dcache-assoc=8 --known-sizeof-char=1 --known-sizeof-void-p=8 --known-sizeof-short=2 --known-sizeof-int=4 --known-sizeof-long=8 --known-sizeof-long-long=8 --known-sizeof-float=4 --known-sizeof-double=8 --known-sizeof-size_t=8 --known-bits-per-byte=8 --known-memcmp-ok=1 --known-sizeof-MPI_Comm=4 --known-sizeof-MPI_Fint=4 --known-mpi-long-double=1 --known-mpi-int64_t=1 --known-mpi-c-double-complex=1 --known-sdot-returns-double=0 --known-snrm2-returns-double=0 --known-has-attribute-aligned=1 --with-cc=mpiicc --with-cxx=mpiicpc --with-fc=0 --with-mpi-lib=/p/system/packages/intel/parallel_studio_xe_2017_update1/compilers_and_libraries_2017.1.132/linux/mpi/lib64/libmpi.so --with-mpi-include=/p/system/packages/intel/parallel_studio_xe_2017_update1/compilers_and_libraries_2017.1.132/linux/mpi/include64 --with-blas-lapack-dir=/p/system/packages/intel/parallel_studio_xe_2017_update1/compilers_and_libraries_2017.1.132/linux/mkl/lib/intel64 --with-64-bit-indices=1 --known-mpi-shared-libraries=1 --with-debugging=1 --with-valgrind=1 --with-valgrind-dir=/p/system/packages/valgrind/3.10.1/gnu --with-x=0 --with-ssl=0 --with-batch=1 --with-shared-libraries=1 --with-mpiexec=srun CFLAGS=\"-fp-model precise -no-fast-transcendentals\" CXXFLAGS=\"-fp-model precise -no-fast-transcendentals\"\n",
			"Intel(R) MPI Library 2017 Update 1 for Linux* OS\n",
			"NetCDF 4.4.1.1 of Jan 30 2017 12:20:18 $.\n",
			"FFTW fftw-3.3.6-pl2.\n",
			"GSL 2.3.\n",
			"PROJ.4 Rel. 4.9.3, 15 August 2016.\n",
			"" ;
```
I don't see major differences in the software dependencies...

**Comment by evangowan** [comment]
Just as a small add, I asked Lu to redo one of her northern hemisphere full glacial cycle runs using to check on the magnitude of this problem (using the standard PISM 0.7 version). The difference between the runs was not significant, it produced nearly identical ice volume and extent at all times. I think she was using 40 km resolution, with restarts every 1000 years.

**Comment by ckhroulev** [comment]
Closing because of the lack of recent activity.
